mbtiles/src/pool.rs

```diff
@@ -26,9 +37,302 @@ impl MbtilesPool {
+          maxzoom: 1
+          minzoom: 0
+          name: Geography Class
+          template: "{{#__location__}}{{/__location__}}{{#__teaser__}}<div style=\"text-align:center;\">\n\n<img src=\"data:image/png;base64,{{flag_png}}\" style=\"-moz-box-shadow:0px 1px 3px #222;-webkit-box-shadow:0px 1px 5px #222;box-shadow:0px 1px 3px #222;\"><br>\n<strong>{{admin}}</strong>\n\n</div>{{/__teaser__}}{{#__full__}}{{/__full__}}"
+          version: 1.0.0
+        "#);
+        assert_eq!(
+            pool.detect_type().await.unwrap(),
+            MbtType::Normalized { hash_view: false }
+        );
+        let t1 = pool.get_tile(0, 0, 0).await.unwrap().unwrap();
+        assert!(!t1.is_empty());
+
+        let (t2, h2) = pool
+            .get_tile_and_hash(MbtType::Normalized { hash_view: false }, 0, 0, 0)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(t2, t1);
+        let expected_hash = Some("1578fdca522831a6435f7795586c235b".to_string());
+        assert_eq!(h2, expected_hash);
+
+        let (t3, h3) = pool
+            .get_tile_and_hash(MbtType::Flat, 0, 0, 0)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(t3, t2);
+        assert_eq!(h3, None);
+        for error_types in [
+            MbtType::FlatWithHash,
+            MbtType::Normalized { hash_view: true },
+        ] {
+            assert!(pool.get_tile_and_hash(error_types, 0, 0, 0).await.is_err());
+        }
+    }
+
+    #[allow(clippy::too_many_lines)]
+    #[tokio::test]
+    async fn test_flat_with_hash() {
+        let pool =
+            MbtilesPool::open_readonly("../tests/fixtures/mbtiles/zoomed_world_cities.mbtiles")
+                .await
+                .unwrap();
+        let metadata = pool.get_metadata().await.unwrap();
+        insta::assert_yaml_snapshot!(metadata,@r#"
+        id: zoomed_world_cities
+        tile_info:
+          format: mvt
+          encoding: gzip
+        layer_type: overlay
+        tilejson:
+          tilejson: 3.0.0
+          tiles: []
+          vector_layers:
+            - id: cities
+              fields:
+                name: String
+              description: ""
+              maxzoom: 6
+              minzoom: 0
```

Comment:
Agree that generating test files might uncover bugs.
Here we are handling this too simplistic for example.

We are trusting that the metadata is actually valid.

Reply:
yep,  and actually generating sqlite files should be very simple - we can just use the `.dump` command in sqlite to export the content of all existing files, delete the originals, clean up the generated dumps, remove any unrelated stuff from them, and use them to generate .mbtiles files during testing - e.g. with some extra `init-mbtiles` just command to set them up in some build dir.